File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *node=head;
        int len=0;
        while(node){
            len++;
            node=node->next;
        }
        return reverseList(head,len,k);

    }

    ListNode* reverseList(ListNode *head,int len,int k){
        if(k>len || !head || !head->next) return head;
        ListNode *dummy=new ListNode(-1);
        int cnt=k;
        dummy->next=head;

        ListNode *prev=dummy,*next=NULL,*node=head;
        while(cnt--){
            next=node->next;
            node->next=prev;
            prev=node;
            node=next;
        }

        dummy->next->next=reverseList(next,len-k,k);
        return prev;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def recursion(self,front,k,n):
        if front==None or n<k: 
            return front 
        curr_node,prev,nextNode,firstcpy = front,None,None,front
        cnt = 0
        while curr_node and cnt!=k:
            nextNode = curr_node.next 
            curr_node.next = prev
            prev=curr_node
            curr_node = nextNode
            cnt+=1
        firstcpy.next = self.recursion(nextNode,k,n-cnt)
        return prev

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # in function reverse node upto 
        # head->next = recursive(nextNode)
        # return lastNode reverse 
        cnt = 0 
        node = head 
        while node: 
            node = node.next 
            cnt+=1 
        return self.recursion(head,k,cnt)